libr/bin/pdb/pdb.c

```diff
@@ -604,466 +617,655 @@ static void finish_pdb_parse(R_PDB *pdb) {
+	ut32 value = type; // cast to unsigned for defined bitwise operations
+	/*   https://llvm.org/docs/PDB/TpiStream.html#type-indices
+        .---------------------------.------.----------.
+        |           Unused          | Mode |   Kind   |
+        '---------------------------'------'----------'
+        |+32                        |+12   |+8        |+0
+	*/
+	// because mode is only number between 0-7, 1 byte is enough
+	return (value & 0x00000000F0000);
+}
 
-	return state;
+static SimpleTypeKind get_simple_type_kind (PDB_SIMPLE_TYPES type) {
+	ut32 value = type; // cast to unsigned for defined bitwise operations
+	/*   https://llvm.org/docs/PDB/TpiStream.html#type-indices
+        .---------------------------.------.----------.
+        |           Unused          | Mode |   Kind   |
+        '---------------------------'------'----------'
+        |+32                        |+12   |+8        |+0
+	*/
+	return (value & 0x00000000000FF);
 }
 
-///////////////////////////////////////////////////////////////////////////////
-/// TODO: rewrite this ?!
-static int build_format_flags(R_PDB *pdb, char *type, int pos, char *res_field, char **name_field) {
-	EStates curr_state;
-	char *tmp = 0;
-	char *name = 0;
-
-	tmp = strtok (type, " ");
-	while (tmp != NULL) {
-		curr_state = convert_to_state (tmp);
-		switch (curr_state) {
-		case eMemberState:
-			break;
-		case ePointerState:
-			if (res_field[pos] == 'p') {
-				return 1;
+/**
+ * @brief Creates the format string and puts it into format
+ * 
+ * @param type_info Information about the member type
+ * @param format buffer for the formatting string
+ * @param names buffer for the member names
+ * @return int -1 if it can't build the format
+ */
+static int build_member_format(STypeInfo *type_info, RStrBuf *format, RStrBuf *names) {
+	r_return_val_if_fail (type_info && format && names && type_info->type_info, -1);
+	// THOUGHT: instead of not doing anything for unknown types I can just skip the bytes
+	// format is 2 chars tops + null terminator
+
+	char *name = "unnamed_field";
+	if (type_info->get_name) {
+		type_info->get_name (type_info, &name);
+	}
+	name = r_str_sanitize_sdb_key (name);
+
+	SType *under_type = NULL;
+	if (type_info->leaf_type == eLF_MEMBER ||
+		type_info->leaf_type == eLF_NESTTYPE) {
+		if (type_info->get_index) {
+			type_info->get_index (type_info, (void **)&under_type);
+		} else {
+			r_warn_if_reached ();
+		}
+	} else if (type_info->leaf_type == eLF_METHOD ||
+		type_info->leaf_type == eLF_ONEMETHOD) {
+		return 0; // skip method member
+	} else {
+		r_warn_if_reached ();
+		return -1;
+	}
+	type_info = &under_type->type_data;
+
+	char *member_format = NULL;
+	char tmp_format[5] = { 0 }; // used as writable format buffer
+	if (type_info->leaf_type == eLF_SIMPLE_TYPE) {
+		SLF_SIMPLE_TYPE *simple_type = type_info->type_info;
+		SimpleTypeMode mode = get_simple_type_mode (simple_type->base_type);
+
+		switch (mode) {
```

Comment:
Maybe extract this huge switch into a separate function called something like `map_type_format` or something like that?

Reply:
Yeah, I've extracted the simple types switch into the separate function and redid the else if chain into a switch